dpytools/http/upload/token_auth.py

```
@@ -0,0 +1,123 @@
+            self.token_creation_time = datetime.now()
+
+            self.auth_token = response_headers["Authorization"]
+            self.id_token = response_headers["ID"]
+        else:
+            err = Exception("Failed to create user tokens")
+            logger.error(
+                "Failed to create user tokens",
+                err,
+                data={
+                    "identity_api_url": self.identity_api_url,
+                    "token_url": token_url,
+                    "response_staus_code": response.status_code,
+                    "response_content": response.content,
+                },
+            )
+            raise err
+
+    def get_auth_header(self) -> Dict[str, str]:
+        """
+        Given a dictionary of params, set the auth header based on the
+        auth mode in use.
+        """
+
+        # Using service account
+        if self.service_token:
+            return {"Authorization": f"Bearer {self.service_token}"}
+
+        # Using user account
+        # If the token is more than 10 minutes old refresh it
+        # https://github.com/ONSdigital/dp-identity-api/blob/develop/swagger.yaml
+        if (datetime.now() - self.token_creation_time) > timedelta(minutes=10):
+            self.refresh_user_token()
+
+        return {"X-Florence-Token": self.auth_token, "ID": self.id_token}
+
+    def refresh_user_token(self):
+        """
+        Refreshes the user token by sending a PUT request to the
+        Identity API's token endpoint with the current refresh token
+        and ID token.
+        """
+        token_refresh_url = f"{self.identity_api_url}/tokens/self"
+        response = self.put(
+            token_refresh_url,
+            json={"Refresh": self.refresh_token, "ID": self.id_token},
+        )
+
+        if response.status_code == 201:
+            self.auth_token = response.headers["Authorization"]
+            self.id_token = response.headers["ID"]
+        else:
+            err = Exception(
```

Could this error be tested? If you patch a made up response to make sure the error triggers correctly.